src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests_Main.cs

```diff
@@ -0,0 +1,401 @@
+            {
+                attribute.VerifyValue(0, TypedConstantKind.Primitive, expectedValue);
+                Assert.Equal($@"MyAttribute(""{expectedValue}"")", attribute.ToString());
+            };
+        }
+
+        [Theory]
+        [InlineData("All", true)]
+        [InlineData("Method", true)]
+        [InlineData("Class", false)]
+        [InlineData("Field", false)]
+        [InlineData("Assembly", false)]
+        [InlineData("Module", false)]
+        public void TestMainAttributesWithSpecificLocation(string location, bool valid)
+        {
+            var source = CreateCompilation($@"
+using System;
+
+[main: My(""one"")]
+Console.WriteLine(""Hello World"");
+
+[AttributeUsage(AttributeTargets.{location}, AllowMultiple = true)]
+public class MyAttribute : Attribute {{ public MyAttribute(string name) {{}} }}
+");
+
+            source.VerifyDiagnostics(valid ? Array.Empty<DiagnosticDescription>() :
+                new[]
+                {
+                    // (4,8): error CS0592: Attribute 'My' is not valid on this declaration type. It is only valid on '...' declarations.
+                    // [main: My("one")]
+                    Diagnostic(ErrorCode.ERR_AttributeOnBadSymbolType, "My").WithArguments("My", location.ToLower()).WithLocation(4, 8)
+                });
+        }
+
+        [Fact]
+        public void TestMainAttributesInScript()
+        {
+            var compilation = CreateSubmission(@"
+using System;
+[main: My(""one"")]
+
+public class C {}
+public class MyAttribute : Attribute { public MyAttribute(string name) {} }",
+            parseOptions: TestOptions.Script.WithLanguageVersion(LanguageVersion.Preview));
+
+            compilation.VerifyDiagnostics(
+                // (3,2): error CS7026: Assembly and module attributes are not allowed in this context
+                // [main: My("one")]
+                Diagnostic(ErrorCode.ERR_GlobalAttributesNotAllowed, "main").WithLocation(3, 2)
+                );
+        }
+
+        private static ImmutableArray<CSharpAttributeData> GetSimpleEntryPointAttributes(ModuleSymbol module) => GetSimpleEntryPoint(module).GetAttributes();
+
+        private static MethodSymbol GetSimpleEntryPoint(ModuleSymbol module) => Assert.Single(GetSimpleEntryPoints(module));
+
+        private static ImmutableArray<MethodSymbol> GetSimpleEntryPoints(ModuleSymbol module)
+        {
+            var program = module.GlobalNamespace.GetMember<NamedTypeSymbol>(WellKnownMemberNames.TopLevelStatementsEntryPointTypeName);
+            var methods = module switch
+            {
+                SourceModuleSymbol => program.GetMembers().Where(m => m is SynthesizedSimpleProgramEntryPointSymbol),
+                _ => program.GetMembers(WellKnownMemberNames.TopLevelStatementsEntryPointMethodName)
+            };
+            return methods.Cast<MethodSymbol>().ToImmutableArray();
+        }
+    }
```

Comment:
Please add some test cases showing local function interactions, something like this scenario:

```cs
[main: My]
void M() {}
```